### DISCUSSION TITLE
How to trigger eviction of LRU cache on my own

### DISCUSSION THREAD

**Discussion by dsaingre** [facebook/CacheLib#296] [Q&A] [answered]
Hi all ! 

As part of a research project, I'm building a LRU-based cache using cachelib. One key need for this project is to be able to manage cache evictions. In particular, I would like to be able to trigger eviction when needed. 

According to the documentation on [MMContainers](https://cachelib.org/docs/Cache_Library_Architecture_Guide/ram_cache_indexing_and_eviction/#mmcontainer), I see that I can use methods like `getEvictionIterator` or `CacheAllocator<CacheTrait>::findEviction(PoolId pid, ClassId cid)` to be able to retrieve the next item to evict. Unfortunately, I believe that those methods are conceived for internal usage by cachelib as I can't access those externally, either from a `CacheAllocator` or a `MemoryPool`. 

For instance, I first thought of doing something like `_gCache->getMMContainer(this->_defaultPool, id).getEvictionIterator();` where `_gCache` is a `std::unique_ptr<facebook::cachelib::LruAllocator> ` (in this case, `getMMContainer` is private).

Am I missing something? Is there any way to get the eviction iterator as a cachelib user, through a public method? 

Please note that in my case, I can't just iterate over the elements contained in the cache. I would like to leverage the LRU eviction mechanism to retrieve the last-recently-used item.

Thanks a lot in advance for any help. If something is not clear, feel free to ask me any precisions.

**Comment by haowu14** [comment]
> Am I missing something? Is there any way to get the eviction iterator as a cachelib user, through a public method?

No. CacheLib does not have a (public) mechanism for users to trigger eviction. In theory, you only need an eviction when you are doing an allocation, and cachelib does that for you when you call cachelib.allocate().

Do you mind specifying what exactly you are trying to achieve? For example, do you have a slightly modified LRU algorithm you want to use in place of cachelib's MMLru's implementation?

**Reply by dsaingre** [reply]
Hi, 

Thanks for your answer. My goal is to be able to trigger cache eviction on command to be able to minimize cache size when needed. I believe I could extend the existing MMLru to add a specific method that would expose the eviction feature. Thanks for confirming that eviction is only triggered internally.

**Reply by haowu14** [reply, marked as answer]
If you want to control what exactly to evict, you can extend the existing MMLru or replace it with your own implementation.
However, depending on what you mean by cache size, it may or maynot work for your purpose. Evicting an item does not free the memory space back to the operating system. If your goal is to free up space to the OS, you can do it via memory monitor: https://cachelib.org/docs/Cache_Library_User_Guides/oom_protection.
Similarly you can modify or use your own implementation there.

**Comment by pbhandar2** [comment]
Closing this discussion as it has been inactive. If this is still relevant, then please create a new discussion post and we will get back to you as soon as we can. Thank you!
